source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name max length must be %d length and NULL terminated",
+            (void *)socket,
+            socket->io_handle.data.fd,
+            AWS_NETWORK_INTERFACE_NAME_MAX);
+        return aws_raise_error(AWS_IO_SOCKET_INVALID_OPTIONS);
+    }
+    if (network_interface_length != 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name is not supported on this platform.",
+            (void *)socket,
+            socket->io_handle.data.fd);
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    struct nw_socket *nw_socket = aws_mem_calloc(alloc, 1, sizeof(struct nw_socket));
+    nw_socket->allocator = alloc;
+
+    socket->allocator = alloc;
+    socket->options = *options;
+    socket->impl = nw_socket;
+    socket->vtable = &s_vtable;
+
+    aws_mutex_init(&nw_socket->synced_data.lock);
+    aws_mutex_init(&nw_socket->synced_state.lock);
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    nw_socket->synced_data.base_socket = socket;
+    aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+    s_set_socket_state(nw_socket, socket, INIT);
+
+    if (s_setup_socket_params(nw_socket, options)) {
+        aws_mem_release(alloc, nw_socket);
+        return AWS_OP_ERR;
+    }
+    aws_ref_count_init(&nw_socket->ref_count, nw_socket, s_socket_impl_destroy);
+    aws_ref_count_init(&nw_socket->internal_ref_count, nw_socket, s_socket_internal_destroy);
+    aws_ref_count_acquire(&nw_socket->ref_count);
+    aws_linked_list_init(&nw_socket->read_queue);
+
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p fd=%d: socket created.", (void *)nw_socket, socket->io_handle.data.fd);
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_client_set_dispatch_queue(struct aws_io_handle *handle, void *queue) {
+    nw_connection_set_queue(handle->data.handle, queue);
+}
+
+static void s_handle_socket_timeout(struct aws_task *task, void *args, aws_task_status status) {
+    (void)task;
+    (void)status;
+
+    struct nw_socket_timeout_args *timeout_args = args;
+    struct nw_socket *nw_socket = timeout_args->nw_socket;
+
+    AWS_LOGF_TRACE(AWS_LS_IO_SOCKET, "task_id=%p: timeout task triggered, evaluating timeouts.", (void *)task);
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    struct aws_socket *socket = nw_socket->synced_data.base_socket;
+    /* successful connection will have nulled out timeout_args->socket */
+    if (!timeout_args->cancelled && socket) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: timed out, shutting down.",
+            (void *)socket,
+            (void *)nw_socket->nw_connection);
+
+        int error_code = AWS_IO_SOCKET_TIMEOUT;
+
+        if (status != AWS_TASK_STATUS_RUN_READY) {
+            error_code = AWS_IO_EVENT_LOOP_SHUTDOWN;
+            // nw_socket->event_loop = NULL;
+        }
+        aws_raise_error(error_code);
```

Comment:
Is this necessary?  We don't raise error (with the callback error code) in callbacks that apply the error code or pass it on.